chain/index/events.go

```diff
@@ -0,0 +1,548 @@
+		tipsetKeyCid, err = si.getTipsetKeyCidByHeight(ctx, f.MinHeight)
+		if err != nil {
+			if err == ErrNotFound {
+				// this means that this is a null round and there exist no events for this epoch
+				return nil
+			}
+
+			return xerrors.Errorf("failed to get tipset key cid by height: %w", err)
+		}
+	default:
+		// Filter doesn't specify a specific tipset
+		return nil
+	}
+
+	// If we couldn't determine a specific tipset, return ErrNotFound
+	if tipsetKeyCid == nil {
+		return ErrNotFound
+	}
+
+	// Check if the determined tipset is indexed
+	exists, err := si.isTipsetIndexed(ctx, tipsetKeyCid)
+	if err != nil {
+		return xerrors.Errorf("failed to check if tipset is indexed: %w", err)
+	}
+
+	if exists {
+		return nil // Tipset is indexed
+	}
+
+	return ErrNotFound // Tipset is not indexed
+}
+
+// getTipsetKeyCidByHeight retrieves the tipset key CID for a given height.
+func (si *SqliteIndexer) getTipsetKeyCidByHeight(ctx context.Context, height abi.ChainEpoch) ([]byte, error) {
+	ts, err := si.cs.GetTipsetByHeight(ctx, height, nil, false)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to get tipset by height: %w", err)
+	}
+
+	if ts.Height() != height {
+		// this means that this is a null round
+		return nil, ErrNotFound
+	}
+
+	return toTipsetKeyCidBytes(ts)
+}
+
+// GetEventsForFilter returns matching events for the given filter
+// Returns nil, nil if the filter has no matching events
+// Returns nil, ErrNotFound if the filter has no matching events and the tipset is not indexed
+// Returns nil, err for all other errors
+func (si *SqliteIndexer) GetEventsForFilter(ctx context.Context, f *EventFilter, excludeReverted bool) ([]*CollectedEvent, error) {
+	getEventsFnc := func(stmt *sql.Stmt, values []any) ([]*CollectedEvent, error) {
+		q, err := stmt.QueryContext(ctx, values...)
+		if err != nil {
+			return nil, xerrors.Errorf("failed to query events: %w", err)
+		}
+		defer func() { _ = q.Close() }()
+
+		var ces []*CollectedEvent
+		var currentID int64 = -1
+		var ce *CollectedEvent
+
+		for q.Next() {
+			select {
+			case <-ctx.Done():
+				return nil, ctx.Err()
+			default:
+			}
+
+			var row struct {
+				id           int64
+				height       uint64
+				tipsetKeyCid []byte
+				emitterID    uint64
+				emitterAddr  []byte
+				eventIndex   int
+				messageCid   []byte
+				messageIndex int
+				reverted     bool
+				flags        []byte
+				key          string
+				codec        uint64
+				value        []byte
+			}
+
+			if err := q.Scan(
+				&row.id,
+				&row.height,
+				&row.tipsetKeyCid,
+				&row.emitterID,
+				&row.emitterAddr,
+				&row.eventIndex,
+				&row.messageCid,
+				&row.messageIndex,
+				&row.reverted,
+				&row.flags,
+				&row.key,
+				&row.codec,
+				&row.value,
+			); err != nil {
+				return nil, xerrors.Errorf("read prefill row: %w", err)
+			}
+
+			if row.id != currentID {
+				if ce != nil {
+					ces = append(ces, ce)
+					ce = nil
```

Comment:
Why are we doing this here, and then again outside of the loop? why don't we just append immediately after making it? even in the MaxResults case we end up appending anyway after the loop. I can't see a good reason for this, can you?

Reply:
@rvagg 

This is to handle the case where we have only one event with one event entry. The loop breaks without the append in that case so we need to take care of it outside.

Reply:
We can't append immediately after making it because event entries are still not in place. 

Basically, we can get it working without that one additional append outside the loop with some refactoring of this code but I really don't want to do it as this code has worked pretty well for reading events with multiple entries so far and solves the problem well.

Reply:
> the event entries are still not in place.

they don't need to be though, you just put them into the last event in the slice

Reply:
```diff
diff --git a/chain/index/events.go b/chain/index/events.go
index 94abb45cb..5ac7fd93f 100644
--- a/chain/index/events.go
+++ b/chain/index/events.go
@@ -335,15 +335,8 @@ func (si *SqliteIndexer) GetEventsForFilter(ctx context.Context, f *EventFilter,
                        // The query will return all entries for all matching events, so we need to keep track
                        // of which event we are dealing with and create a new one each time we see a new id
                        if row.id != currentID {
-                               if ce != nil {
-                                       ces = append(ces, ce)
-                                       ce = nil
-                                       // Unfortunately we can't easily incorporate the max results limit into the query due to the
-                                       // unpredictable number of rows caused by joins
-                                       // Break here to stop collecting rows
-                                       if f.MaxResults > 0 && len(ces) >= f.MaxResults {
-                                               break
-                                       }
+                               if f.MaxResults > 0 && len(ces) >= f.MaxResults {
+                                       break
                                }
 
                                currentID = row.id
@@ -353,6 +346,7 @@ func (si *SqliteIndexer) GetEventsForFilter(ctx context.Context, f *EventFilter,
                                        Height:   abi.ChainEpoch(row.height),
                                        MsgIdx:   row.messageIndex,
                                }
+                               ces = append(ces, ce)
 
                                if row.emitterAddr == nil {
                                        ce.EmitterAddr, err = address.NewIDAddress(row.emitterID)
@@ -392,10 +386,6 @@ func (si *SqliteIndexer) GetEventsForFilter(ctx context.Context, f *EventFilter,
                        })
                }
 
-               if ce != nil {
-                       ces = append(ces, ce)
-               }
-
                if len(ces) == 0 {
                        return nil, nil
                }
```

Reply:
But what does this change get us ?

Reply:
Are you okay to punt it ? Have tested the existing query logic  thoroughly for twats and it works well. Don't want to mess with it unless there is a strong reason to.

Reply:
simplified maintenance is my argument for such a change - the fewer places you repeat the same thing, the less liklihood there is of a future maintainer adjusting something and not realising they need to adjust it in two places

Reply:
your call, but this is _smell_ to me

Reply:
You are right -> fixed this. It's done.